### PR TITLE
Add installer build/test to the runtime.yml pipeline

### PR DESCRIPTION
**Pull request by dagood**
Adds installer build and test to `runtime.yml`, consuming live artifacts. Preserves the old standalone installer build + test pipeline using cached bits for now, but if it breaks with https://github.com/dotnet/runtime/pull/494 as expected I won't try to repair it with this PR.

Old WIP PR: https://github.com/dotnet/runtime/pull/529.

/cc @trylek @safern @jashook

### CODE REVIEW

File: eng/pipelines/common/platform-matrix.yml
Comment (safern) [inline eng/pipelines/common/platform-matrix.yml]: This is being fixed by https://github.com/dotnet/runtime/pull/642 also. 

Should we instead close that PR? We might want to run the tests @jashook mentioned there.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (safern) [inline eng/pipelines/installer/jobs/base-job.yml]: This could be moved to an `_IsOfficialBuild` common variable and we could add a variable template that is imported in runtime.yml that way we avoid duplication and these hard to read conditions. Something like: https://github.com/dotnet/runtime/blob/master/eng/pipelines/libraries/variables.yml#L9

File: eng/pipelines/installer/jobs/osx-build.yml
Comment (safern) [inline eng/pipelines/installer/jobs/osx-build.yml]: Just to learn, what does: `>-` do?

Comment (trylek) [reply eng/pipelines/installer/jobs/osx-build.yml]: I'm seeing this for the first time too...

Comment (dagood) [reply eng/pipelines/installer/jobs/osx-build.yml]: `>` replaces newlines with spaces (vs. `|` which keeps newlines), `-` strips the newline at the end. http://yaml-multiline.info/ is a nice resource for this, and it has links to the spec.

File: eng/pipelines/libraries/build-job.yml
Comment (safern) [inline eng/pipelines/libraries/build-job.yml]: I guess this will need to be rebased if: https://github.com/dotnet/runtime/pull/494 is merged first.

File: eng/pipelines/installer/jobs/osx-build.yml
Comment (safern) [reply eng/pipelines/installer/jobs/osx-build.yml]: Thanks, pretty useful 😄

File: eng/pipelines/common/clone-checkout-bundle-step.yml
Comment (safern) [inline eng/pipelines/common/clone-checkout-bundle-step.yml]: Should Libraries and CoreCLR jobs be updated to use this template? Should just be a couple of files: https://github.com/dotnet/runtime/search?q=Checkout_bundle&unscoped_q=Checkout_bundle

File: eng/pipelines/common/platform-matrix.yml
Comment (dagood) [reply eng/pipelines/common/platform-matrix.yml]: About why it's here, I imagine we need the artifacts from the platform for this PR. (@jkoritzinsky can confirm.)

Are you referring to running the coreclr outerloop?

I think it would be easiest to drive the other PR to merge because it's more self-contained and probably easier to verify. Then the diff can be removed here.

Comment (jkoritzinsky) [reply eng/pipelines/common/platform-matrix.yml]: I originally added this just so the job naming for the installer jobs was correct. I didn't realize that this was blocking the ARM testing at that time. We should be able to temporarily remove this and allow #642 to add it.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (dagood) [reply eng/pipelines/installer/jobs/base-job.yml]: Interesting... I never got to grips with variable templates (or using variable values in template expressions), this is a good starting point. I'll give this a shot, thanks.

File: eng/pipelines/libraries/build-job.yml
Comment (dagood) [reply eng/pipelines/libraries/build-job.yml]: Yeah, I think the MSBuild changes are probably going to be the bigger conflicts though. I'm not expecting to be able to merge them in particularly quick succession. 😄

File: eng/pipelines/common/clone-checkout-bundle-step.yml
Comment (dagood) [reply eng/pipelines/common/clone-checkout-bundle-step.yml]: Yeah, I initially waved off on it to avoid adding unnecessary changes while everyone's working on the yaml in various ways, but I'll give it a shot.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (trylek) [reply eng/pipelines/installer/jobs/base-job.yml]: Just please remember the variable templates can only be imported at the "top level" i.e. at the pipeline YAML level, not from the lower-level templates, that's the current limitation AFAIK. Like - it would be super-logical to import a variable template section into one of the CoreCLR template - it just doesn't work. I initially tried to mimic variables in templates by splitting platform-matrix in three different templates, then I rolled it back because it plain sucked. Santi currently worked it around by using explicit nested templates (the xplat-setup stuff). I believe this is just poor design on Azure Pipelines' part.

Comment (dagood) [reply eng/pipelines/installer/jobs/base-job.yml]: Yeah... weird limits on templates. I don't see any fundamental reason you shouldn't be able to use them anywhere you want. (Whether it gets messy is up to the author!) Really confusing.

I guess this specific weirdness is because *pipeline* variables are "statically defined", but *job* variables aren't... but the only reference I find to statically defined variables is https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops so I can't tell if that's what's intended:

> In a parse time expression, you have access to `parameters` and statically-defined `variables`.

There's feedback filed on a few pages about this being confusing. 😛

Comment (safern) [reply eng/pipelines/installer/jobs/base-job.yml]: Yeah. It would have to be imported in runtime.yml and then you can do `variables['IsOfficialBuild']`.... That's what we're doing for the libraries build now.

**Comment by dagood** [comment]
In the last CI run, I somehow got AzDO to give me a good runtime pipeline build of Installer using live bits. All the platforms ran but macOS, which was just stalling due to the ongoing AzDO issues.

@jkoritzinsky @ViktorHofer @safern @trylek PTAL

File: eng/pipelines/common/variables.yml
Comment (safern) [inline eng/pipelines/common/variables.yml]: I might be missing something, but where is this imported?

Comment (dagood) [reply eng/pipelines/common/variables.yml]: Well, it's imported in Libraries' `eng\pipelines\libraries\variables.yml`. But yeah, not used by Installer yet (or anywhere else) because we don't have a def set up to do official builds yet. Shuffling this out of the CoreFX-specific vars just lets me switch `${{ and(ne(variables['System.TeamProject'], 'public'), notIn(variables['Build.Reason'], 'PullRequest')) }}` to `isOfficialBuild` in this PR with clear intent where it *will* come from once I work on an official build.

**Comment by dagood** [comment]
CI is all green except `runtime-coreclr (CoreCLR Pri0 Test Run R2R Windows_NT x86 checked)`, which timed out trying to send (get?) Helix results. I hit retry.

Would appreciate more reviews. 🙂

**Review by trylek — APPROVED** [review]
OMG, sorry for not signing off, your change looks awesome to me!

**Comment by jkoritzinsky** [comment]
A lot of the Helix queues are very full (the Windows queue that PRs use currently has a queue depth of 2.3k work items), so we might want to wait for the queue to drain before re-running any more if it times out again.

File: eng/pipelines/installer/installer-matrix.yml
Comment (jkoritzinsky) [reply eng/pipelines/installer/installer-matrix.yml]: It might be possible to make skipTests = crossBuild. Based on the current file, that seems correct.

File: eng/pipelines/runtime.yml
Comment (jkoritzinsky) [inline eng/pipelines/runtime.yml]: I think we should invert this to instead specify the platforms we want to build for, not the platforms we want to skip. We should do this in a similar manner to the `platform-matrix.yml` file is done.

File: src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets
Comment (jkoritzinsky) [inline src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: Why is the `NuGetPackageId` metadata required? Is that related to the support in Microsoft.DotNet.Build.Tasks.SharedFramework.Sdk for crossgenning the framework?

File: src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets
Comment (jkoritzinsky) [inline src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: What part of this target did you have to override/patch?

File: eng/pipelines/common/xplat-setup.yml
Comment (dagood) [reply eng/pipelines/common/xplat-setup.yml]: We have them, it's just a no-op right now on Win a few lines up:

```
- name: setScriptToEchoAndFailOnNonZero
  value: ''
```

Don't know if it's possible to do this (or necessary) in batch.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (dagood) [reply eng/pipelines/installer/jobs/base-job.yml]: > I originally chickened out of Santi's suggestion to put this logic in xplat-setup (and you're not using it, after all)

I actually am using xplat-setup, since I use platform-matrix... or am I misunderstanding?

It makes sense to me to have it in xplat-setup, since the bulk of the transfer is setting up variables, and I think this is the only reasonable mechanism we have to share agent-local variable setup.

File: eng/pipelines/common/variables.yml
Comment (safern) [reply eng/pipelines/common/variables.yml]: I see. I guess this file will be imported in runtime.yml for the official builds work under a global variable section just as `eng\pipelines\libraries\.azure-ci.yml` does for it's variables template.

We can leave that as-side for the official build effort.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (jkoritzinsky) [reply eng/pipelines/installer/jobs/base-job.yml]: > > I originally chickened out of Santi's suggestion to put this logic in xplat-setup (and you're not using it, after all)
> 
> I actually am using xplat-setup, since I use platform-matrix... or am I misunderstanding?

We really need to update our docs on how these templates flow. I'll open a tracking issue for that.

File: src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets
Comment (dagood) [reply src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: I believe it was a batch somewhere that assumed all items had `NuGetPackageId` defined. I wouldn't expect it to change any concrete results if the assumption were fixed instead, but this was easy and matches how we're adding this data to the CoreFX items.

File: src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets
Comment (dagood) [reply src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: See below:

```
      <!-- !!!BEGIN CHANGE Persist existing TargetPath. -->
      <RidSpecificFilesToPackage Condition="'%(RidSpecificFilesToPackage.TargetPath)' == ''">
      <!-- !!!END CHANGE -->
```

Although I guess I neglected mentioning the original: [`<RidSpecificFilesToPackage>`](https://github.com/dotnet/arcade/blob/d2d8d9f5c14e353acc7dfcf69c5fa5a8c368af40/src/Microsoft.DotNet.Build.Tasks.SharedFramework.Sdk/targets/framework.dependency.targets#L20).

File: eng/pipelines/installer/jobs/base-job.yml
Comment (safern) [reply eng/pipelines/installer/jobs/base-job.yml]: The reasoning was to have all this `*ArtifactName` variables and `LegName` defined in `xplat-setup` so that all templates can benefit from it and just use them without having to actually define it in multiple places. Currently each partition defined this on it's base template.

However the change is not trivial as the right parameters need to be flown to the right templates in order to be able to access the values we need on `xplat-setup`. So this can be a follow up clean-up change. 

I still have to do clean-up on `libraries` template which I left from my previous PR to unblock work, so I can put it as part of my change.

File: eng/pipelines/runtime.yml
Comment (safern) [reply eng/pipelines/runtime.yml]: Anyway I think it makes sense to add a comment so that whenever we expand the full matrix for all partitions we don't forget adding any platform.

**Review by safern — APPROVED** [review]

File: eng/pipelines/installer/jobs/base-job.yml
Comment (trylek) [reply eng/pipelines/installer/jobs/base-job.yml]: Please let's make sure we don't get deadlocked by these considerations for now - it's probably up to me to apologize for bringing them up in the first place. I'll be the happiest person in the world to clean this up but it's strictly ordered after standing up the official build and producing signing artifacts so please let's not block on that for now. I'll be happy to play with pipeline cleanups over / after the EOY holidays ;-).

**Comment by dagood** [comment]
I don't think any of the convos are blockers--I'd like to get this merged to unblock and work on them incrementally. I've filed https://github.com/dotnet/runtime/issues/749 to track the issues. I'll file PRs for smaller ones really quick when I have gaps in my work on signing/publishing. Any objection (while that CoreCLR leg is finishing up)?

File: src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets
Comment (jkoritzinsky) [reply src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: The paths without the change should work as-is. Were they not working correctly?

Comment (jkoritzinsky) [reply src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: In fact, the paths should have been the same as what they're overridden to here.

Comment (dagood) [reply src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: The cross-target paths get reset by that, and need to be preserved:

https://github.com/dotnet/runtime/blob/4e23cd3deabebf6c1f269379eeb01dbadc4be22e/src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets#L47-L49

I think they used to work by being added after `GetFilesFromPackageResolve` rather than before.

Comment (jkoritzinsky) [reply src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: Good catch!

Comment (dagood) [reply src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets]: Well, I did have the benefit of seeing the build fail before I did this. 😄 When the target paths are the same between cross and target archs, the zip gets duplicate entries which fails an unzip step later.

**Pull request merged**
